spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/annotation/SleuthSpanCreateBeanPostProcessor.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2013-2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.annotation;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Method;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.aop.aspectj.annotation.AspectJProxyFactory;
+import org.springframework.aop.support.AopUtils;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanPostProcessor;
+
+/**
+ * This bean post processor analyzes whether the bean is eligible to
+ * be wrapped with an aspect. That will be the case only if there are
+ * {@link NewSpan} annotations either on the class or the interface
+ * methods.
+ *
+ * @author Christian Schwerdtfeger
+ * @since 1.2.0
+ */
+class SleuthSpanCreateBeanPostProcessor implements BeanPostProcessor {
```

Comment:
Aplying aspects through `BeanPostProcessor` is not good approach, in case of circular dependency `BeanCurrentlyInCreationException` will be thrown:
```
@Service
public class Service2 {

    @Autowired
    private Service1 service1;

    @NewSpan
    public void serve2() {
    }
}

@Service
public class Service1 {

    @Autowired
    private Service2 service2;

    @NewSpan
    public void serve1() {
    }
}
```

leads to 
> BeanCurrentlyInCreationException: Error creating bean with name 'service1': Bean with name 'service1' has been injected into other beans [service2] in its raw version as part of a circular reference, but has eventually been wrapped. This means that said other beans do not use the final version of the bean. This is often the result of over-eager type matching - consider using 'getBeanNamesOfType' with the 'allowEagerInit' flag turned off, for example.

Creation aspect bean in this case much better. Also I tried to place annotation on the class and it doesn't working.
I didn't understand the issue with pointcut for interfaces with annotated methods, but I tried to register aspect with following signature:
```
@Pointcut("execution(public * *(..)) && @annotation(newSpan)")
public void annotatedMethod(NewSpan newSpan) {
}

@Pointcut("execution(public * *(..)) && @within(newSpan)")
public void annotatedClass(NewSpan newSpan) {
}

@Around("annotatedMethod(newSpan) || annotatedClass(newSpan)")
public Object instrumentOnMethodAnnotation(ProceedingJoinPoint pjp, NewSpan newSpan) throws Throwable {
	Method method = getMethod(pjp);
        Assert.notNull(method, "method can't be null inside aspecting pointcut");
	Span span = null;
	try {
		span = this.spanCreator.createSpan(pjp, newSpan);
		return pjp.proceed();
	} finally {
		if (span != null) {
			this.tracer.close(span);
		}
	}
}
```
It works fine for both classes and methods, also we don't need `SleuthAnnotationUtils` to find annotation.